### PR TITLE
Count growing segment number to estimate cpu usage

### LINKED ISSUE #17819: [Enhancement]: Count growing segment number to estimate cpu usage
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What would you like to be added?

Currently, when estimating the CPU load of the search task, the number of growing segments is hardcode 5.
When there are many growing segments, this will cause many concurrent tasks to compete for CPU.
The number of real growing segments should be counted.

### Why is this needed?

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by czs007**
fix #17819 
Signed-off-by: zhenshan.cao <zhenshan.cao@zilliz.com>

### CODE REVIEW

File: internal/querynode/meta_replica.go [hunk 1]
```diff
@@ -477,6 +481,31 @@ func (replica *metaReplica) getSegmentIDs(partitionID UniqueID, segType segmentT
 	return replica.getSegmentIDsPrivate(partitionID, segType)
 }
 
+// getGrowingSegmentNumByVChannel returns number of segments which virtual channel is vChannel
+// if partitionIDs is empty, it means that filtering by partitionIDs is not required.
+func (replica *metaReplica) getGrowingSegmentNumByVChannel(vChannel Channel, partitionIDs []UniqueID) int {
```
Comment (xiaofan-luan) [inline internal/querynode/meta_replica.go:486]: I thought we should extend getSegmentIDsByVChannel to support get Streaming segmentIds?
There is no reason to add a new interface for get count

**Comment by xige-16** [comment]
releated issue #17513

File: internal/querynode/meta_replica.go [hunk 2]
```diff
@@ -478,25 +479,40 @@ func (replica *metaReplica) getSegmentIDs(partitionID UniqueID, segType segmentT
...
+	for _, segment := range segments {
 		if segment.vChannelID == vChannel {
-			segmentIDsTmp = append(segmentIDsTmp, segment.ID())
+			if filterPartition {
+				partitionID := segment.partitionID
+				for _, partID := range partitionIDs {
+					if partitionID == partID {
+						ret = append(ret, segment.ID())
+						break
+					}
```
Comment (sunby) [inline internal/querynode/meta_replica.go:508]: you can construct a map to filter partition, do not traverse all partition ids per segment

File: internal/querynode/task_search.go [hunk 3]
```diff
@@ -179,10 +179,20 @@ func (s *searchTask) Notify(err error) {
...
+		channel := s.req.GetDmlChannel()
+		segIDs, err := s.QS.metaReplica.getSegmentIDsByVChannel(partitionIDs, channel, segmentTypeGrowing)
+		if err != nil {
+			log.Error("searchTask estimateCPUUsage", zap.Error(err))
+		}
+		segmentNum := len(segIDs)
+		if segmentNum <= 0 {
+			segmentNum = 1
+		}
+		s.cpu = int32(s.NQ) * int32(segmentNum)
```
Comment (sunby) [inline internal/querynode/task_search.go:192]: int32(s.NQ * segmentNum)

File: internal/querynode/scheduler_policy.go [hunk 4]
```diff
@@ -4,13 +4,13 @@ import (
 	"container/list"
 )
 
-type scheduleReadTaskPolicy func(sqTasks *list.List, targetUsage int32) ([]readTask, int32)
+type scheduleReadTaskPolicy func(sqTasks *list.List, targetUsage int32, maxNum int32) ([]readTask, int32)
 
-func defaultScheduleReadPolicy(sqTasks *list.List, targetUsage int32) ([]readTask, int32) {
+func defaultScheduleReadPolicy(sqTasks *list.List, targetUsage int32, maxNum int32) ([]readTask, int32) {
```
Comment (sunby) [inline internal/querynode/scheduler_policy.go:9]: Add some ut

File: internal/querynode/task_search.go [hunk 5]
```diff
@@ -179,10 +179,20 @@ func (s *searchTask) Notify(err error) {
...
 	} else if s.DataScope == querypb.DataScope_Historical {
 		segmentNum := int64(len(s.req.GetSegmentIDs()))
-		s.cpu = int32(s.NQ * segmentNum / 2)
+		s.cpu = int32(s.NQ * segmentNum)
```
Comment (xiaofan-luan) [inline internal/querynode/task_search.go:195]: need a param here to config cpu usage ratio

**Pull request merged**